### DISCUSSION TITLE
A question about custom controls for WinUI 3

### DISCUSSION THREAD

**Discussion by AFunDog** [microsoft/WindowsAppSDK#4541] [Q&A]
I am a beginner.
I want to know how to package custom controls into the Nuget package so that I can use them in other projects.

**Comment by michael-hawker** [comment]
FYI @Karl-Bridge-Microsoft @alvinashcraft 

Related:
- https://github.com/microsoft/WindowsAppSDK/issues/3546
- https://learn.microsoft.com/windows/apps/develop/platform/csharp-winrt/create-winrt-component-winui-cswinrt
- https://learn.microsoft.com/en-us/nuget/create-packages/overview-and-workflow (this probably most relevant but still all UWP based)

There's a few projects like the [Windows Community Toolkit](https://aka.ms/toolkit/windows) which do this. Generally, if you just create a new Class Library (WinUI 3 in Desktop) project and include the GenerateLibraryLayout msbuild property in your csproj, I think that gets you most of the way for basic scenarios (for extra assets there's the issue above). It's then generally best if you want to create a sharable library to use Templated Controls over User Controls as those allow for greater flexibility and performance for others.
